### PR TITLE
feat(electron): add connectOverCDP function to playwright._electron

### PR DESCRIPTION
**Pull request by ZFail**
Added connectOverCDP functionality to playwright._electron, requested at https://github.com/microsoft/playwright/issues/10369

### CODE REVIEW

**Comment by ZFail** [comment]
@microsoft-github-policy-service agree

**Comment by mastrzyz** [comment]
Love This!

File: packages/playwright-core/src/server/electron/electron.ts [hunk 1]
```diff
@@ -114,6 +120,28 @@ export class ElectronApplication extends SdkObject {
   }
 }
 
+async function urlToNodeWSEndpoint(progress: Progress, endpointURL: string) {
+  if (endpointURL.startsWith('ws'))
+    return endpointURL;
+  progress.log(`<ws preparing> retrieving websocket url from ${endpointURL}`);
+  const httpURL = endpointURL.endsWith('/') ? `${endpointURL}json/` : `${endpointURL}/json/`;
```
Comment (macabeus) [inline packages/playwright-core/src/server/electron/electron.ts:127]: Please, remove the last `/`.
```suggestion
  const httpURL = endpointURL.endsWith('/') ? `${endpointURL}json` : `${endpointURL}/json`;
```

For some reason, the `json/` endpoint here returns 404: 
image

But without the last slash works:
image

**Comment by macabeus** [comment]
I tried to use this branch, but I'm always receiving the following error when trying to evaluate anything on the page:
```
JavaScriptErrorInEvaluate: ReferenceError: process is not defined
```

The following commands returned it:
```js
const result = await electronApp.evaluate(() => {
  return 1 + 2;
})
```
```js
const window = await electronApp.firstWindow();
console.log(await window.title());
```

Despite that, `electronApp.windows()` returned the right amount of pages.

**Comment by ZFail** [comment]
> ```js
> const result = await electronApp.evaluate(() => {
>   return 1 + 2;
> })
> ```

on my machine this code work and returns 3, i`m add this to my test https://github.com/microsoft/playwright/pull/20094/files#diff-25a018cac50f8f5e3a9f2a7691a61391be2e5995e74d9cc03a058be43163a795R230

**Comment by macabeus** [comment]
I'm working on macOS

**Comment by ZFail** [comment]
> I'm working on macOS

on macos, the test also passes, can you please run the test with the command `npx playwright test electron-app.spec.ts -g "^(.*[\\\/])?electron-app.spec.ts.should connect to external electron$"`

**Comment by macabeus** [comment]
The test passes. Perhaps it's failing here because of the electron application itself

**Comment by jiby-aurum** [comment]
@ZFail just tried it, works well, thanks for the work.

**Comment by jiby-aurum** [comment]
@dgozman @aslushnikov is this something the core team would be willing to land. Would love to use it from upstream

**Comment by pavelfeldman** [comment]
I don't think we are comfortable committing to the over-CDP support for Electron. Playwright needs to connect to both Node and the browser in a specific order and we are deferring the electron activation to make sure there are no internal races. Connecting over CDP requires us to add similar dance around the races that we would not want to maintain.

**Pull request closed without merging**